### PR TITLE
`ember new` with blueprints from npm

### PR DESCRIPTION
**Pull request by tomdale**
This PR adds support for installing blueprints from npm. Previously, we supported installing blueprints from the local filesystem or via an explicit `.git` URL.

This feature works by looking at the blueprint name and only attempting to install the blueprint from npm if:

1. The blueprint is a valid npm package name.
2. No blueprint with that name was found locally.

This ends up being nicer from a developer ergonomics perspective because I can just say something like "Type `ember new my-app -b pwa-blueprint`" instead of "Type `ember new my-app -b https://github.com/tomdale/pwa-blueprint.git`".

The feature is wrapped in the `NPM_BLUEPRINTS` experiment flag so hopefully it can be iterated on without too much risk if people think this is a good overall direction.

### CODE REVIEW

File: lib/tasks/install-blueprint.js [hunk 1]
```diff
@@ -38,22 +46,100 @@ class InstallBlueprintTask extends Task {
...
+  let pkgPath = path.join(tempPath, 'node_modules', blueprintName);
+
+  return execa('npm', ['install', blueprintName], { cwd: tempPath })
+    .catch(err => throwNpmError(err))
+    .then(() => pkgPath);
+}
+
+const NOT_FOUND_REGEXP = /npm ERR! 404 {2}'(\S+)' is not in the npm registry/;
+
+function throwNpmError(err) {
```
Comment (stefanpenner) [inline lib/tasks/install-blueprint.js]: can we unit test this function?

File: lib/tasks/install-blueprint.js [hunk 2]
```diff
@@ -38,22 +46,100 @@ class InstallBlueprintTask extends Task {
...
+// repository so it has all of its dependencies.
+function fromGit(blueprintName, tempPath) {
+  return execa('git', ['clone', blueprintName, tempPath])
+    .then(() => execa('npm', ['install'], { cwd: tempPath }))
+    .then(() => tempPath);
+}
+
+// Install the blueprint with npm. We skip running `npm install` inside like the
+// git version because npm handles this for us automatically.
+function fromNpm(blueprintName, tempPath) {
```
Comment (stefanpenner) [inline lib/tasks/install-blueprint.js]: can we unit test this function?

File: lib/tasks/install-blueprint.js [hunk 3]
```diff
@@ -38,22 +46,100 @@ class InstallBlueprintTask extends Task {
...
+    return lookupBlueprint(blueprintName, blueprintPaths)
+      .then(blueprint => blueprint.install(installOptions));
+  }
+}
+
+// Given a blueprint name and set of paths to search, looks up the specified
+// blueprint. If it detects a git repository URL, it will clone the repository
+// and use that as the blueprint. If it detects an npm package name, it will
+// install the package and use that as the blueprint.
+function lookupBlueprint(blueprintName, paths) {
```
Comment (stefanpenner) [inline lib/tasks/install-blueprint.js]: can we unit test this function?

**Comment by Turbo87** [comment]
@tomdale @stefanpenner I can take over on this if you want me to

**Comment by tomdale** [comment]
@Turbo87 Seems legit, I don't think I'll be able to get to it today.

**Comment by Turbo87** [comment]
@tomdale I'm wondering if we should use a `npm:` prefix for this (e.g. `ember new my-app -b npm:pwa-blueprint`). mixing up blueprint names, git urls and npm package names without some way to distinguish them feels a little confusing to me.

**Comment by stefanpenner** [comment]
> @tomdale I'm wondering if we should use a npm: prefix for this (e.g. ember new my-app -b npm:pwa-blueprint). mixing up blueprint names, git urls and npm package names without some way to distinguish them feels a little confusing to me.

I like that! @tomdale any downside?

**Comment by tomdale** [comment]
I don't love the idea of a prefix, because the motivation is to have something as concise as possible for people to type.

In practice I don't think there's actually that much ambiguity (at least, specifically for the `ember new` case):

1. No ambiguity in git URLs as we require them to be very explicit, including a `.git` prefix at the end.
2. The set of `new`-compatible blueprints is very bounded (just `app` and `addon`).
3. Everything else is an npm package.

I think there's a more worrying ambiguity case with `ember generate`, where the likelihood of collisions is higher. As a compromise, how would we feel about scoping npm-sourced blueprints to just the `ember new` case?

I'd also be open to a different flag that took an `npm install`-compatible package identifier (e.g. `ember-cli/template` for GitHub, `gist:ember-cli/template` for Gists, `bitbucket:foo/bar`, etc.)

**Comment by rwjblue** [comment]
> I don't love the idea of a prefix, because the motivation is to have something as concise as possible for people to type.

I agree with @tomdale here, I'd prefer to avoid the prefix in what seems like the default case (npm packages) for `ember new` or `ember addon`.

> As a compromise, how would we feel about scoping npm-sourced blueprints to just the ember new case?

I think this would be fine, and as you mentioned we only support `app` and `addon` as blueprint names outside of a project context anyways so there is little worry of collisions...

**Pull request closed without merging**